File: src/main/java/no/nav/fo/veilarbregistrering/registrering/bruker/HentRegistreringService.java
package no.nav.fo.veilarbregistrering.registrering.bruker;

import no.nav.fo.veilarbregistrering.bruker.Bruker;
import no.nav.fo.veilarbregistrering.orgenhet.Enhetnr;
import no.nav.fo.veilarbregistrering.orgenhet.NavEnhet;
import no.nav.fo.veilarbregistrering.orgenhet.Norg2Gateway;
import no.nav.fo.veilarbregistrering.profilering.Profilering;
import no.nav.fo.veilarbregistrering.profilering.ProfileringRepository;
import no.nav.fo.veilarbregistrering.registrering.BrukerRegistreringType;
import no.nav.fo.veilarbregistrering.registrering.manuell.ManuellRegistrering;
import no.nav.fo.veilarbregistrering.registrering.manuell.ManuellRegistreringRepository;
import no.nav.fo.veilarbregistrering.registrering.manuell.Veileder;
import no.nav.fo.veilarbregistrering.registrering.formidling.Status;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import static no.nav.fo.veilarbregistrering.registrering.bruker.Resending.kanResendes;

public class HentRegistreringService {

    private static final Logger LOG = LoggerFactory.getLogger(HentRegistreringService.class);

    private final BrukerRegistreringRepository brukerRegistreringRepository;
    private final SykmeldtRegistreringRepository sykmeldtRegistreringRepository;
    private final ProfileringRepository profileringRepository;
    private final ManuellRegistreringRepository manuellRegistreringRepository;
    private final Norg2Gateway norg2Gateway;

    public HentRegistreringService(
            BrukerRegistreringRepository brukerRegistreringRepository,
            SykmeldtRegistreringRepository sykmeldtRegistreringRepository,
            ProfileringRepository profileringRepository,
            ManuellRegistreringRepository manuellRegistreringRepository,
            Norg2Gateway norg2Gateway) {
        this.brukerRegistreringRepository = brukerRegistreringRepository;
        this.sykmeldtRegistreringRepository = sykmeldtRegistreringRepository;
        this.profileringRepository = profileringRepository;
        this.manuellRegistreringRepository = manuellRegistreringRepository;
        this.norg2Gateway = norg2Gateway;
    }

    public OrdinaerBrukerRegistrering hentOrdinaerBrukerRegistrering(Bruker bruker) {
        return hentOrdinaerBrukerRegistrering(bruker,
                List.of(Status.OVERFORT_ARENA, Status.PUBLISERT_KAFKA, Status.OPPRINNELIG_OPPRETTET_UTEN_TILSTAND));
    }

    public OrdinaerBrukerRegistrering hentIgangsattOrdinaerBrukerRegistrering(Bruker bruker) {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = hentOrdinaerBrukerRegistrering(bruker,
                List.of(Status.DOD_UTVANDRET_ELLER_FORSVUNNET, Status.MANGLER_ARBEIDSTILLATELSE));
        return kanResendes(ordinaerBrukerRegistrering) ? ordinaerBrukerRegistrering : null;
    }

    private OrdinaerBrukerRegistrering hentOrdinaerBrukerRegistrering(Bruker bruker, List<Status> status) {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = brukerRegistreringRepository
                .finnOrdinaerBrukerregistreringForAktorIdOgTilstand(bruker.getAktorId(), status)
                .stream()
                .findFirst()
                .orElse(null);

        if (ordinaerBrukerRegistrering == null) {
            return null;
        }

        Veileder veileder = hentManuellRegistreringVeileder(
                ordinaerBrukerRegistrering.getId(), ordinaerBrukerRegistrering.hentType());
        ordinaerBrukerRegistrering.setManueltRegistrertAv(veileder);

        Profilering profilering = profileringRepository.hentProfileringForId(
                ordinaerBrukerRegistrering.getId());
        ordinaerBrukerRegistrering.setProfilering(profilering);

        return ordinaerBrukerRegistrering;
    }

    public SykmeldtRegistrering hentSykmeldtRegistrering(Bruker bruker) {
        SykmeldtRegistrering sykmeldtBrukerRegistrering = sykmeldtRegistreringRepository
                .hentSykmeldtregistreringForAktorId(bruker.getAktorId());

        if (sykmeldtBrukerRegistrering == null) {
            return null;
        }
        Veileder veileder = hentManuellRegistreringVeileder(
                sykmeldtBrukerRegistrering.getId(), sykmeldtBrukerRegistrering.hentType());
        sykmeldtBrukerRegistrering.setManueltRegistrertAv(veileder);

        return sykmeldtBrukerRegistrering;
    }

    public Veileder hentManuellRegistreringVeileder(long registreringId, BrukerRegistreringType brukerRegistreringType) {
        ManuellRegistrering registrering = manuellRegistreringRepository
                .hentManuellRegistrering(registreringId, brukerRegistreringType);

        if (registrering == null) {
            return null;
        }

        Optional<NavEnhet> enhet = finnEnhet(Enhetnr.Companion.of(registrering.getVeilederEnhetId()));

        return new Veileder()
                .setEnhet(enhet.orElse(null))
                .setIdent(registrering.getVeilederIdent());
    }

    Optional<NavEnhet> finnEnhet(Enhetnr enhetId) {
        try {
            Map<Enhetnr, NavEnhet> enhetnrNavEnhetMap = norg2Gateway.hentAlleEnheter();
            NavEnhet navEnhet = enhetnrNavEnhetMap.get(enhetId);
            return Optional.ofNullable(navEnhet);
        } catch (Exception e) {
            LOG.error("Feil ved henting av NAV-enheter fra den nye Organisasjonsenhet-tjenesten.", e);
            return Optional.empty();
        }
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/registrering/bruker/HentRegistreringService.kt
package no.nav.fo.veilarbregistrering.registrering.bruker

import no.nav.fo.veilarbregistrering.bruker.Bruker
import no.nav.fo.veilarbregistrering.log.logger
import no.nav.fo.veilarbregistrering.metrics.Events
import no.nav.fo.veilarbregistrering.metrics.JaNei
import no.nav.fo.veilarbregistrering.metrics.MetricsService
import no.nav.fo.veilarbregistrering.orgenhet.Enhetnr
import no.nav.fo.veilarbregistrering.orgenhet.NavEnhet
import no.nav.fo.veilarbregistrering.orgenhet.Norg2Gateway
import no.nav.fo.veilarbregistrering.profilering.ProfileringRepository
import no.nav.fo.veilarbregistrering.registrering.BrukerRegistreringType
import no.nav.fo.veilarbregistrering.registrering.bruker.resources.BrukerRegistreringWrapper
import no.nav.fo.veilarbregistrering.registrering.bruker.resources.BrukerRegistreringWrapperFactory
import no.nav.fo.veilarbregistrering.registrering.formidling.Status.*
import no.nav.fo.veilarbregistrering.registrering.ordinaer.BrukerRegistreringRepository
import no.nav.fo.veilarbregistrering.registrering.ordinaer.OrdinaerBrukerRegistrering
import no.nav.fo.veilarbregistrering.registrering.sykmeldt.SykmeldtRegistrering
import no.nav.fo.veilarbregistrering.registrering.sykmeldt.SykmeldtRegistreringRepository
import no.nav.fo.veilarbregistrering.registrering.veileder.ManuellRegistreringRepository
import no.nav.fo.veilarbregistrering.registrering.veileder.Veileder

class HentRegistreringService(
    private val brukerRegistreringRepository: BrukerRegistreringRepository,
    private val sykmeldtRegistreringRepository: SykmeldtRegistreringRepository,
    private val profileringRepository: ProfileringRepository,
    private val manuellRegistreringRepository: ManuellRegistreringRepository,
    private val norg2Gateway: Norg2Gateway,
    private val metricsService: MetricsService
) {
    fun hentBrukerregistreringUtenMetrics(bruker: Bruker): BrukerRegistreringWrapper? {
        val ordinaerBrukerRegistrering = hentOrdinaerBrukerRegistrering(bruker)
        val sykmeldtBrukerRegistrering = hentSykmeldtRegistrering(bruker)
        return BrukerRegistreringWrapperFactory.create(ordinaerBrukerRegistrering, sykmeldtBrukerRegistrering)
    }

    fun hentBrukerregistrering(bruker: Bruker): BrukerRegistreringWrapper? {
        val brukerRegistreringWrapper = hentBrukerregistreringUtenMetrics(bruker)
        if (brukerRegistreringWrapper == null) {
            logger.info("Bruker ble ikke funnet i databasen.")
            metricsService.registrer(Events.HENT_BRUKERREGISTRERING_BRUKER_FUNNET, JaNei.NEI)
        }
        metricsService.registrer(Events.HENT_BRUKERREGISTRERING_BRUKER_FUNNET, JaNei.JA)
        return brukerRegistreringWrapper
    }

    fun hentOrdinaerBrukerRegistrering(bruker: Bruker): OrdinaerBrukerRegistrering? {
        val ordinaerBrukerRegistrering = brukerRegistreringRepository
            .finnOrdinaerBrukerregistreringForAktorIdOgTilstand(
                bruker.aktorId,
                listOf(OVERFORT_ARENA, PUBLISERT_KAFKA, OPPRINNELIG_OPPRETTET_UTEN_TILSTAND))
            .firstOrNull()
            ?: return null

        val veileder = hentManuellRegistreringVeileder(
            ordinaerBrukerRegistrering.id, ordinaerBrukerRegistrering.hentType()
        )
        ordinaerBrukerRegistrering.manueltRegistrertAv = veileder
        val profilering = profileringRepository.hentProfileringForId(ordinaerBrukerRegistrering.id)
        return ordinaerBrukerRegistrering.med(profilering)
    }

    fun hentSykmeldtRegistrering(bruker: Bruker): SykmeldtRegistrering? {
        val sykmeldtBrukerRegistrering = sykmeldtRegistreringRepository
            .hentSykmeldtregistreringForAktorId(bruker.aktorId) ?: return null
        val veileder = hentManuellRegistreringVeileder(
            sykmeldtBrukerRegistrering.id, sykmeldtBrukerRegistrering.hentType()
        )
        sykmeldtBrukerRegistrering.manueltRegistrertAv = veileder
        return sykmeldtBrukerRegistrering
    }

    private fun hentManuellRegistreringVeileder(
        registreringId: Long,
        brukerRegistreringType: BrukerRegistreringType
    ): Veileder? {
        val (_, _, _, veilederIdent, veilederEnhetId) = manuellRegistreringRepository
            .hentManuellRegistrering(registreringId, brukerRegistreringType) ?: return null
        val enhet = finnEnhet(Enhetnr(veilederEnhetId))

        if (veilederEnhetId == null) {
            logger.warn("veilederEnhetId er null for registreringId: $registreringId")
        }

        logger.info("Henter NAV-enhet: ${enhet} for enhetId: $veilederEnhetId")

        return Veileder(veilederIdent, enhet)
    }

    fun finnEnhet(enhetId: Enhetnr): NavEnhet? {
        return try {
            norg2Gateway.hentAlleEnheter()[enhetId]
        } catch (e: Exception) {
            logger.error("Får ikke populert registrering med NAV-enhet pga. feil mot NORG2 - fortsetter uten.", e)
            null
        }
    }
}
